### ISSUE TITLE
Published package versions are inconsistent

### ISSUE THREAD

**Issue by TeaDrivenDev** [JaCraig/Craig-s-Utility-Library#42]
Most NuGet packages for the library are currently at version 4.0.304. `CraigsUtilityLibrary-FileFormats` is still at 3.4.199, though, and requires the `DataTypes` and `IO` packages to be at exactly 3.4.199 as well. That means those cannot be updated to their current versions as long as the `FileFormats` package is installed.

Would it be possible to re-publish the `FileFormats` package as version 4.0.304, so the rest of the packages can be used in their most recent versions?

**Comment by JaCraig** [comment]
No. Couple of reasons for that:

1. the change from 3 to 4 was huge for a number of areas of the library. It was basically a rewrite. Doing an in place upgrade for your app is going to lead to not fun times depending on what you were using.
2. The FileFormats namespace was actually rolled into IO. That's why FileFormats stops at 3.4.199. So in 4.0 on, it's just a portion of the IO package.
3. Depending on the file format you are using, it may not exist in the 4.x version of the library. A bunch of the more unpopular formats were killed off on the web, etc. As such I didn't see a point in bringing them forward.

You see even though the library has a ton of downloads, a bunch of forks, stars, etc. I never really wrote it for other people. These were random classes, extension methods, etc. that I wrote for my side projects and I just needed a spot to keep them around. So going from version 1 to 2 to 3 to even 4, the library would completely change as my coding style changed, new side projects had different requirements, etc. and all of that made me change what I added to the library and how I used it. Similarly if I hadn't touched something in a really long time, I just dropped it from the next major version of the library. I didn't want to support something if I had no intention of ever using it again. So things like my LDAP code that was in 3.X is not present in 4.X. SQLHelper didn't make the jump either. Both of them were used fairly frequently by other people but I had mostly stopped using them. People can keep using them as they're still in 3.X but by 4.X, I didn't have the need. 

4.X is what my need was at the time I designed/implemented it. However Microsoft went and threw a wrench in that library by bringing out .Net Core and the .Net Standard. With .Net Standard and .Net Core, I basically have to rethink the library again. To add to that bit of drama (and .Net Core is nothing but drama for the majority of its existence thus far), large sections of the code base can't make the jump at all until .Net Standard 2.0 is finalized (and even then some parts are never going to make the jump because they're tied to the Windows platform's APIs). It has gotten to the point where breaking up the library completely makes the most sense and it's what I'm doing right now. For instance:

* IoC has become [Canister](https://github.com/JaCraig/Canister).
* Media has become [Structure.Sketching](https://github.com/JaCraig/Structure.Sketching)
* AOP is now [Aspectus](https://github.com/JaCraig/Aspectus), which is actually an old library that I've brought back.
* The File IO portion of IO is now [FileCurator](https://github.com/JaCraig/FileCurator)
* Compression is now [Corset](https://github.com/JaCraig/Corset)
* All of the random extension methods, etc. from DataTypes is now [Big Book of Data Types](https://github.com/JaCraig/BigBookOfDataTypes)
* Logging portion of IO is now [WoodCutter](https://github.com/JaCraig/Woodcutter)
* Serialization portion of IO is now [SerialBox](https://github.com/JaCraig/SerialBox)
* Validation is now [Valkyrie](https://github.com/JaCraig/Valkyrie)
* SQLHelper is brought back and now it is its own library as just [SQLHelper](https://github.com/JaCraig/SQLHelper)
* The database modeling code from the ORM is now its own thing as [Data.Modeler](https://github.com/JaCraig/Data.Modeler)
* Random is now [Mirage](https://github.com/JaCraig/Mirage)
* And the ORM code is still a work in progress but is at [Inflatable](https://github.com/JaCraig/Inflatable).

And that doesn't include large portions of the code that's currently in 4.0. So there should be about 10 more by the time things are said and done, including one for file formats. The up side is, of course, that I get to redesign portions of code that I'm not that happy with. For instance SQLHelper was simplified quite a bit, the IoC container now supports open generics, Mirage has a bunch of added capabilities over the current Random code and is much cleaner, Structure.Sketching is a cross platform image manipulation library with a ton of extra filters over the current Media namespace, etc. And each library is mostly independent of the others. There are some exceptions to that (Canister is pretty much used by all of them, BigBook is tied to the AOP library, etc.) but they're mostly self contained now.

So TL;DR: Other than bug fixes for 4.X, this library is pretty much stable and not changing. Bringing portions of 3.X into 4.X, if they're missing, doesn't really make sense for me but I may bring them back as their separate library. If I do they will be heavily modified probably.  But for the FileFormats namespace you should be good if you use IO in 4.X.

**Comment by TeaDrivenDev** [comment]
I understand; thanks for clarifying that. I'll try and find out what we're actually using and what it would make sense to upgrade to.

Thanks
